File: backend/app/models/product.py
from __future__ import annotations

from decimal import Decimal
from uuid import UUID

from sqlalchemy import Boolean, ForeignKey, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.database import Base
from app.models.base import TimestampMixin, UUIDMixin


class Product(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "products"

    __table_args__ = {
        "schema": "store"
    }

    category_id: Mapped[UUID] = mapped_column(
        ForeignKey("store.categories.id"),
        nullable=False,
        index=True,
    )

    name: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
    )

    slug: Mapped[str] = mapped_column(
        String(150),
        nullable=False,
        unique=True,
        index=True,
    )

    short_description: Mapped[str | None] = mapped_column(
        String(200),
        nullable=True,
    )

    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    price: Mapped[Decimal] = mapped_column(
        Numeric(10, 2),
        nullable=False,
    )

    discount_price: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 2),
        nullable=True,
    )

    is_active: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
    )

    category: Mapped["Category"] = relationship(
        back_populates="products",
    )
    
    images: Mapped[list["ProductImage"]] = relationship(
        back_populates="product",
        cascade="all, delete-orphan",
        order_by="ProductImage.display_order",
    )

File: frontend/src/types/product.ts
import { Category } from "./category";

export interface ProductImage {
  public_id: string;
  url: string; // Cloudinary URL
  stored_filename: string;
  original_filename: string;
  mime_type: string;
  file_size: number;
  alt_text: string | null;
  display_order: number;
  is_primary: boolean;
  created_at: string;
  updated_at: string;
}

export interface Product {
  public_id: string;
  name: string;
  slug: string;
  price: number;
  discount_price: number | null;
  short_description: string | null;
  description: string | null;
  is_active: boolean;
  category: Category;
  images: ProductImage[];
  primary_image: string | null; // computed from images
  created_at: string;
  updated_at: string;
}